proc/threads.c

```diff
@@ -1275,138 +1402,309 @@ int threads_sigpost(process_t *process, thread_t *thread, int sig)
+
+	hal_spinlockClear(&current->spinlock, &sc);
+	lock->owner = current;
+
+	return EOK;
+}
 
-	if (!hal_started())
+
+int proc_lockTry(lock_t *lock)
+{
+	spinlock_ctx_t sc;
+	int err;
+
+	if (!hal_started()) {
 		return -EINVAL;
+	}
 
 	hal_spinlockSet(&lock->spinlock, &sc);
-	if (lock->v == 0)
-		err = -EBUSY;
 
-	lock->v = 0;
+	err = _proc_lockTry(lock);
+
 	hal_spinlockClear(&lock->spinlock, &sc);
 
 	return err;
 }
 
 
-int _proc_lockClear(lock_t *lock)
+static int _proc_lockUnlock(lock_t *lock)
 {
-	lock->v = 1;
-	if (lock->queue == NULL || lock->queue == (void *)-1)
+	spinlock_ctx_t sc;
+	int ret = 0;
+
+	hal_spinlockSet(&lock->owner->spinlock, &sc);
+
+	LIST_REMOVE(&lock->owner->locks, lock);
+
+	hal_spinlockClear(&lock->owner->spinlock, &sc);
+	lock->owner = NULL;
+
+	if (lock->queue != NULL) {
+		hal_spinlockSet(&threads_common.spinlock, &sc);
+
+		_proc_threadDequeue(lock->queue);
+
+		hal_spinlockClear(&threads_common.spinlock, &sc);
+		ret = 1;
+	}
+
+	return ret;
+}
+
+
+static void proc_lockUnlock(lock_t *lock)
+{
+	spinlock_ctx_t sc;
+
+	hal_spinlockSet(&lock->spinlock, &sc);
+
+	if (_proc_lockUnlock(lock)) {
+		hal_cpuReschedule(&lock->spinlock, &sc);
+		return;
+	}
+
+	hal_spinlockClear(&lock->spinlock, &sc);
+}
+
+
+static int _proc_lockClear(lock_t *lock)
+{
+	thread_t *current = proc_current();
+	spinlock_ctx_t sc;
+
+	if (lock->owner != current) {
+		return -EPERM;
+	}
+
+	if (!_proc_lockUnlock(lock)) {
 		return 0;
+	}
+
+	hal_spinlockSet(&threads_common.spinlock, &sc);
+
+	current->priority = _proc_threadGetPriority(current);
+
+	hal_spinlockClear(&threads_common.spinlock, &sc);
 
-	proc_threadWakeup(&lock->queue);
 	return 1;
 }
 
 
 int proc_lockClear(lock_t *lock)
 {
 	spinlock_ctx_t sc;
+	int err;
 
-	if (!hal_started())
+	if (!hal_started()) {
 		return -EINVAL;
+	}
 
 	hal_spinlockSet(&lock->spinlock, &sc);
-	if (_proc_lockClear(lock))
+
+	err = _proc_lockClear(lock);
+	if (err > 0) {
 		hal_cpuReschedule(&lock->spinlock, &sc);
-	else
-		hal_spinlockClear(&lock->spinlock, &sc);
+		return EOK;
+	}
 
-	return EOK;
+	hal_spinlockClear(&lock->spinlock, &sc);
+
+	return err;
 }
 
 
 int proc_lockSet2(lock_t *l1, lock_t *l2)
 {
 	int err;
 
-	if ((err = proc_lockSet(l1)) < 0)
+	err = proc_lockSet(l1);
+	if (err < 0) {
 		return err;
+	}
 
 	while (proc_lockTry(l2) < 0) {
 		proc_lockClear(l1);
-		if ((err = proc_lockSet(l2)) < 0)
+		err = proc_lockSet(l2);
+		if (err < 0) {
 			return err;
+		}
 		swap(l1, l2);
 	}
+
 	return EOK;
 }
 
 
 int proc_lockWait(thread_t **queue, lock_t *lock, time_t timeout)
 {
-	int err;
 	spinlock_ctx_t sc;
+	int err;
+
+	if (!hal_started()) {
+		return -EINVAL;
+	}
 
 	hal_spinlockSet(&lock->spinlock, &sc);
-	_proc_lockClear(lock);
-	if ((err = proc_threadWaitEx(queue, &lock->spinlock, timeout, 1, &sc)) != -EINTR)
-		_proc_lockSet(lock, 0, &sc);
+
+	err = _proc_lockClear(lock);
+	if (err >= 0) {
+		err = proc_threadWaitEx(queue, &lock->spinlock, timeout, 1, &sc);
+		if (err == EOK) {
```

Comment:
Uh, no, we need to take the lock if -ETIME. Don't take the lock only in -EINTR case.

Reply:
Done.